Upgrades to typehints and PHP 8.4 method signatures

niden opened the pull request:
- ~Changed minimum version to PHP 8.1 << Please advise if this is acceptable~ (review)
- ~Added version to composer.json~ (review)
- Upgraded phpstan and phpunit (composer)
- Added github actions workflow to run phpstan and the tests
- ~Upgraded phpunit.xml.dist for the new phpunit version~ (review)
- ~Added docblocks for all methods~ (out of scope)
- Added phpstan types for logEntry, connection instances etc.
- Adjusted tests
- Changed implicitly nullable variables to accept null also (PHP 8.4)
- Changed logic in `LoggerStatement::bindValue()` to check with `empty()` instead of `null`
- Changed `PersistentStatement` usages of `func_get_args()` to passed parameters or vadiadic ones (soothing stan)
- Changed `PersistentStatement::errorInfo()` to return only array (parent interface changed)
- Changed `PersistentStatement::fetch()` to use `PDO::FETCH_DEFAULT` instead of `null`. 
- Changed `PersistentStatement::fetchAll()` to return `array` instead of `array|false` and use `PDO::FETCH_DEFAULT` instead of `PDO::FETCH_BOTH`. 

For this one we can also use `#[\ReturnTypeWillChange]` for now, since the return type `PersistentLoggedStatement::fetchAll()` is not covariant with `PDOStatement::fetchAll()`.  As for the mode change, the relevant test was failing without that change. We can either adjust the test or leave that change there.

Please advise what needs to be changed.

Thanks

niden committed:
adjusted .gitignore

niden committed:
converted phpunit to the new format

niden committed:
upgrading packages

niden committed:
phpstan adjustments

niden committed:
adjusting tests

niden committed:
adding ci.yml for actions (tests/quality)

niden committed:
trying a different workflow to run quality and tests

niden committed:
reverting added code for default

niden committed:
corrections for PHP 8.4

niden committed:
added checkout command for tests

niden committed:
reverting change

niden committed:
docblock correction

niden commented on src/ConnectionLocator.php:
```diff
@@ -77,18 +146,27 @@ public function setWriteFactory(
                 $this->defaultFactory,
-                static::DEFAULT
+                $type
             );
```
This was the only way I could find to soothe stan.

pmjones replied:
I'm good with that!

niden committed:
reverting bindValue to mixed

niden committed:
removed extra docblocks (out of scope)

niden committed:
removing more docblocks

niden committed:
removing more docblocks to reduce noise in the PR

froschdesign commented:
I suggest also introducing a [backward compatibility check](https://github.com/Roave/BackwardCompatibilityCheck).

niden commented:
> I suggest also introducing a [backward compatibility check](https://github.com/Roave/BackwardCompatibilityCheck).

This is what the tool reported. 

```
[BC] CHANGED: Default parameter value for parameter $fetch_style of Atlas\Pdo\PersistentLoggedStatement#fetch()
    changed from NULL to 0
[BC] CHANGED: The parameter $fetch_style of Atlas\Pdo\PersistentLoggedStatement#fetch() changed from int|null 
    to a non-contravariant int
[BC] CHANGED: The parameter $fetch_style of Atlas\Pdo\PersistentLoggedStatement#fetch() changed from int|null 
    to int
[BC] CHANGED: Default parameter value for parameter $fetch_style of Atlas\Pdo\PersistentLoggedStatement#fetchAll()
     changed from 4 to 0
[BC] CHANGED: The return type of Atlas\Pdo\PersistentLoggedStatement#fetchAll() changed from array|false 
    to array
[BC] CHANGED: The parameter $ctor_args of Atlas\Pdo\PersistentLoggedStatement#fetchObject() changed from 
    array|null to a non-contravariant array
[BC] CHANGED: The parameter $ctor_args of Atlas\Pdo\PersistentLoggedStatement#fetchObject() changed from 
    array|null to array
[BC] CHANGED: The return type of Atlas\Pdo\PersistentLoggedStatement#errorInfo() changed from array|null 
    to array
```

froschdesign commented:
I think my suggestion was too vague: the backward compatibility check should be added to the CI pipeline.

> This is what the tool reported.

BC breaks are [not allowed in a minor version](https://semver.org/#spec-item-8) therefore these changes must be reverted or a new major version is required.

pmjones commented on composer.json:
```diff
@@ -1,6 +1,7 @@
 {
     "name": "atlas/pdo",
     "type": "library",
+    "version": "2.0.0",
```
Composer recommends against adding the version; it will get picked up automatically from the branch name.

niden replied:
I saw the warning when composer run the update... :) 

Removed it. Thanks

pmjones commented on composer.json:
```diff
@@ -17,7 +18,8 @@
     ],
     "require": {
-        "php": "^8.0"
+        "php": "^8.1",
```
If we bump the PHP version, that's likely a new major version release for the library -- any chance of keeping it down at 8.0, even if it means losing some language niceties?

niden replied:
Reverted to 8.0 min.

pmjones commented on src/LoggedStatement.php:
```diff
@@ -13,16 +13,23 @@
+     * @param logEntry $logEntry
+     */
     protected function __construct(
         protected mixed /* callable */ $queryLogger,
```
With the `@param callable` in the docblock, you can remove the mid-line comment here.

pmjones commented on src/LoggedStatement.php:
```diff
@@ -35,9 +42,10 @@ public function bindValue(
         $result = parent::bindValue($parameter, $value, $dataType);
 
-        if ($result && $this->logEntry !== null) {
+        if ($result && !empty($this->logEntry)) {
```
I don't think empty() is quite the same thing as `!== null` -- can you say why the change?

niden replied:
The `logEntry` is defined as an array in the constructor so it does not accept nulls, as such it will not be null here. We can change it if you wish but then line 29 will complain (constructor).

pmjones commented on src/PersistentLoggedStatement.php:
```diff
@@ -31,11 +41,15 @@ static public function new(
     private array $logEntry;
 
-    /* Atttributes */
+    /* Attributes */
```
😄

pmjones commented on src/PersistentLoggedStatement.php:
```diff
@@ -107,37 +137,46 @@ public function execute(array $inputParameters = null) : bool
 
     public function fetch(
-        int $fetch_style = null,
+        int $fetch_style = PDO::FETCH_DEFAULT,
```
This looks like a BC break ... ?

niden replied:
It is. If I run phpstan for 8.0 it complains:

_Parameter #1 $mode of method PDOStatement::fetch() expects int, int|null given._

For PHP 8.4 it complains for the lack of `?int` also. 

I am not sure how to tackle this. Perhaps an exception in phpstan config?

pmjones commented on src/PersistentLoggedStatement.php:
```diff
@@ -107,37 +137,46 @@ public function execute(array $inputParameters = null) : bool
 
     public function fetchAll(
-        int $fetch_style = PDO::FETCH_BOTH,
+        int $fetch_style = PDO::FETCH_DEFAULT,
```
This also looks like a BC break ... ?

niden replied:
I was following the tests on this. If `FETCH_BOTH` is set then the ` Atlas\Pdo\PersistentLoggedStatementTest::testFetchAll` test fails (because there are more data than what the test checks for). We can certainly adjust the test and leave this as it was.

pmjones commented on src/PersistentLoggedStatement.php:
```diff
@@ -107,37 +137,46 @@ public function execute(array $inputParameters = null) : bool
     public function fetchObject(
         ?string $class_name = 'stdClass',
-        ?array $ctor_args = []
+        array $ctor_args = []
```
Also a BC break ... ?

niden replied:
For this one again phpstan complains:

_Parameter #2 $constructorArgs of method PDOStatement::fetchObject() expects array<mixed>, array|null given._

We can leave the original types and add an exception to phpstan.

pmjones commented on src/PersistentLoggedStatement.php:
```diff
@@ -164,7 +203,7 @@ public function errorCode() : ?string
     }
 
-    public function errorInfo() : ?array
+    public function errorInfo() : array
```
BC break ... ?

niden replied:
This relates to the comment I left in the description of the PR.

_Return type array|null of method Atlas\Pdo\PersistentLoggedStatement::errorInfo() is not covariant with tentative return type array of method PDOStatement::errorInfo()._

Either this change or `#[\ReturnTypeWillChange]`

pmjones reviewed (changes requested):
Hey @niden thank you so much for this work! This is the first pass at a review; I expect to send a couple more before we're through.

FWIW there's no need to (as far as project standards are concerned) to do alignment of equals, or docblock params, etc.

niden committed:
rename stan type logEntryType; added php 8.0 in the run; removed version from composer changed min to 8.0

niden committed:
changing to 8.0 minimum version; adjusting composer and tests

niden committed:
using empty for logEntry

niden commented:
@pmjones Thanks for the review. Some replies above when you get a chance.

pmjones commented on src/Connection.php:
```diff
@@ -21,9 +21,27 @@
  * @method string quote(mixed $string, int $parameterType = PDO::PARAM_STR)
  * @method mixed setAttribute(int $attribute, mixed $value)
+ *
+ * @phpstan-type dsnArgs array{
```
For non-classlike types, I prefer to see lower-snake-case, with the type as a suffix if possible. That helps to distinguish them from classlike types, and follows the PHPStan naming style pretty closely (though we cannot use dashes in userland custom types.) So, `dsn_args_array` or something like that.

pmjones commented on src/Connection.php:
```diff
@@ -21,9 +21,27 @@
+ *      options?: array<int, mixed>
+ *  }
+ *
+ * @phpstan-type logEntryType array{
```
Likewise, maybe something like `log_entry_array`.

pmjones commented on src/Connection.php:
```diff
@@ -49,11 +68,14 @@ static public function factory(mixed ...$args) : callable
     public function __construct(protected PDO $pdo)
     {
-        $this->persistent = $this->pdo->getAttribute(PDO::ATTR_PERSISTENT);
+        $this->persistent = (bool)$this->pdo->getAttribute(PDO::ATTR_PERSISTENT);
```
Nit: space between (bool) and $this.

pmjones commented on src/Connection.php:
```diff
@@ -164,6 +190,7 @@ protected function performBind(
         }
 
         if (! is_array($args)) {
+            /** @var int|string $name */
```
Might be better to put this as `@param int|string $name` on the method?

pmjones commented on src/Connection.php:
```diff
@@ -253,6 +301,7 @@ public function fetchObjects(
     ) : array|false
     {
         $sth = $this->perform($statement, $values);
+        /** @var array<array-key, callable|int|string> $args */
```
Would this @var be better as @param on the method signature?

pmjones commented on src/Connection.php:
```diff
@@ -262,7 +311,10 @@ public function fetchOne(
     {
         $sth = $this->perform($statement, $values);
-        return $sth->fetch(PDO::FETCH_ASSOC);
+        /** @var array<array-key, mixed> $result */
```
Might this be good as a custom type, e.g. `fetch_assoc_array mixed[]` and the the @return type would be `fetch_assoc_array|false` ?

pmjones commented on src/Connection.php:
```diff
@@ -306,6 +358,7 @@ public function yieldUnique(
         $sth = $this->perform($statement, $values);
 
         while ($row = $sth->fetch(PDO::FETCH_UNIQUE)) {
+            /** @var array<string, mixed> $row */
```
Maybe another opportunity for a custom type.

pmjones commented on src/ConnectionLocator.php:
```diff
@@ -10,12 +10,29 @@
+    /**
+     * @var string
+     */
     public const DEFAULT = 'DEFAULT';
```
I don't think it's a BC break to add `string` typehints on these ... though it might not be 8.0 compatible. Thoughts?

niden replied:
We cannot add `public const string DEFAULT` because it is not supported in php 8.0.

Sadly `public const /* string */ DEFAULT = 'DEFAULT';` does not work. Phpstan still complains about it.

pmjones commented on src/ConnectionLocator.php:
```diff
@@ -79,14 +110,17 @@ public function setWriteFactory(
     {
-        if ($this->instances[static::DEFAULT] === null) {
-            $this->instances[static::DEFAULT] = $this->newConnection(
+        /** @var 'DEFAULT' $type */
```
This might not be necessary with a switch to `const string DEFAULT`.

pmjones commented on src/ConnectionLocator.php:
```diff
@@ -127,7 +161,12 @@ protected function getConnection(
 
         if (! empty($this->instances[$type])) {
-            return reset($this->instances[$type]);
+            /** @var array<string, array<string, Connection>> $instances */
```
Is this a repetition of the connectionStore type?

niden replied:
Somewhat. I have created a new type to handle this and the store array.

pmjones reviewed (changes requested):
Second pass, still more to go.

In some other projects I have started added a `{Project}CustomTypes` interface or final class, and then import everything from that. Something to consider here, I think.

niden committed:
added custom types class; corrections after review

niden committed:
connection entry type; minor adjustments based on review

niden commented:
@pmjones thanks for the second pass. I fixed most of it, explanations on the ones I could not.
